packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1756,6 +1764,38 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
     }
 
     public setConnectionState(connected: boolean, clientId?: string) {
+        if (connected === false && this.delayConnectedP !== undefined) {
+            this.delayConnectedP.canceled = true;
+            this.delayConnectedP = undefined;
+            this.mc.logger.sendTelemetryEvent({
+                eventName: "UnsuccessfulConnectedTransition",
+            });
+            // Don't propagate "disconnected" event because we didn't propagate the previous "connected" event
+            return;
+        }
+
+        // If attachment blobs were added while disconnected, we need to delay
+        // propagation of the "connected" event until we have uploaded them to
+        // ensure we don't submit ops referencing a blob that has not been uploaded
+        const connecting = connected && !this._connected && !this.deltaManager.readOnlyInfo.readonly;
+        if (connecting && this.blobManager.hasPendingUploads) {
+            const canceled = false;
+            const promise = this.blobManager.onConnected().then(() => {
+                this.delayConnectedP = undefined;
+                if (!canceled && !this.disposed) {
+                    this.setConnectionStateCore(connected, clientId);
+                }
+            }, (reason) => {
+                this.closeFn(reason);
```

Comment:
wouldn't onConnect's promise be rejected if connection get interrupted in the middle.  Would we `close`  the container in that case?